dvc/tree/dropbox.py

```diff
@@ -0,0 +1,466 @@
+REFRESH_TOKEN = "DROPBOX_REFRESH_TOKEN"
+
+
+Auth = namedtuple("Auth", ["client", "creds"])
+
+
+class DropboxCredProvider(metaclass=ABCMeta):
+    DROPBOX_APP_KEY = ""
+    DROPBOX_APP_SECRET = ""
+
+    @abstractmethod
+    def can_auth(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def _gather_auth(self):
+        raise NotImplementedError
+
+    def auth(self):
+        import dropbox
+
+        config = self._gather_auth()
+        if config[EXPIRATION]:
+            config[EXPIRATION] = datetime.datetime.fromisoformat(
+                config[EXPIRATION]
+            )
+        else:
+            config[EXPIRATION] = datetime.datetime.now()
+        dbx = dropbox.Dropbox(
+            oauth2_access_token=config[ACCESS_TOKEN],
+            oauth2_access_token_expiration=config[EXPIRATION],
+            oauth2_refresh_token=config[REFRESH_TOKEN],
+            app_key=self.DROPBOX_APP_KEY,
+            app_secret=self.DROPBOX_APP_SECRET,
+        )
+        dbx.check_and_refresh_access_token()
+        try:
+            dbx.check_user()
+        except dropbox.exceptions.AuthError as ex:
+            raise DvcException(
+                "To use Dropbox account, you need to login.\n"
+                "If you already did, your credentials may be outdated.\n"
+                "Learn more at {}".format(
+                    format_link("https://man.dvc.org/remote/modify")
+                )
+            ) from ex
+        return Auth(DropboxWrapper(dbx), config)
+
+    @abstractmethod
+    def save(self, creds):
+        raise NotImplementedError
+
+
+class CliCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return True
+
+    def _gather_auth(self):
+        import dropbox
+
+        flow = dropbox.DropboxOAuth2FlowNoRedirect(
+            consumer_key=self.DROPBOX_APP_KEY,
+            consumer_secret=self.DROPBOX_APP_SECRET,
+            token_access_type="offline",
+        )
+        authorize_url = flow.start()
+        print("Go to the following link in your browser:")
+        print()
+        print("    " + authorize_url)
+        print()
+        code = input("Enter verification code: ").strip()
+        try:
+            auth = flow.finish(code)
+        except dropbox.oauth.NotApprovedException as ex:
+            raise DvcException(
+                "To use Dropbox remote, you need to approve the DVC "
+                "application to access your Dropbox account.\n"
+                "Learn more at {}".format(
+                    format_link("https://man.dvc.org/remote/modify")
+                )
+            ) from ex
+        return {
+            ACCESS_TOKEN: auth.access_token,
+            EXPIRATION: auth.expires_at.isoformat(),
+            REFRESH_TOKEN: auth.refresh_token,
+        }
+
+    def save(self, creds):
+        pass
+
+
+class EnvCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return bool(os.environ.get(REFRESH_TOKEN, False))
```

Comment:
I'll refer to my comment above. You actually need only the refresh token, as it allows you to get the access token anytime. Refresh token is not ephemeral one-time token just for a single refresh, you can use it as many time as you want.

Reply:
interesting, thanks! 

Thinking from the user perspective I still think it's better try to this type of auth if we detected any of those vars (and fail if they are not enough, e.g. REFRESH_TOKEN is not set). Just to be extra explicit. To avoid situations you set some vars, but it still uses file w/o any explanation. WDYT?

Reply:
I thought of two approaches:

- requiring just the refresh token - that seems reasonable if it's the only must-have element
- requiring everything - in the case they will shorten life of refresh token or make it usable just once

It's your decision, I can adjust the code as needed.

Reply:
So, there are two different things (correct me if I wrong)

1. Trigger - some factor that defines that we go the path of a specific auth (and don't try others)
2. Validation - check if set of params is enough to go that path

So, in this specific case (to my mind):

1. Can be any of the env vars present. It's better from the user perspective, we are extra explicit.
2. Can check existence of the refresh token. Or also allow both keys to be present (I'm not sure if it makes sense at all)?


Btw, how would the actual workflow with env vars would look like? Where do users get them in the first place?